experimental/JuLie/partitions.jl

```diff
@@ -497,93 +563,177 @@ end
   end
-
+  
+  #This will be the list of all partitions found.
+  P = Partition{T}[] 
+
+  # Now, we get to the partb algorithm. This is a hell of an algorithm and the
+  # published code has several issues.
+  # The original ALGOL 60 code is electronically available at
+  # https://gist.github.com/ulthiel/99de02994fc31fe614586ed0c930f744.
+  # First, there were some issues with termination and indices for the arrays 
+  # x,y,ii getting out of bounds. We had to introduce some additional checks 
+  # and breaks to take care of this. Some initial fixing was done by T. Schmit.
+  # An example showing the problem is 17, 3, [1,4], [1,4].
+
+  # Initialize variables
   r = length(v)
   j = 1
   k = mu[1]
   ll = v[1]
-  x = zeros(Int8, n+1)
-  y = zeros(Int8, n+1)
-  ii = zeros(Int8, n)
-  i_1 = 0
-  P = Partition{T}[]
+  x = zeros(T, n)
+  y = zeros(T, n)
+  ii = zeros(T, n)
 
-  num = 0
+  # The algorithm has three goto labels b1, b2, b3.
+  # b3 terminates the algorithm.
+  # We introduce bools to indicate the jumps.
   gotob2 = false
   gotob1 = true
 
-  for i = n:-1:1
+  # The first step in the algorithm is to initialize the arrays x and y
+  # for the backtrack search. 
+  # This fills the array x from right with the values from v from the left 
+  # (the smallest) up to their specified multiplicity.
+  # If this is larger than m, there cannot be a partition.
+  for i=n:-1:1
     x[i] = ll
     y[i] = ll
     k = k - 1
     m = m - ll
     if k == 0
       if j == r
-        return P
+        # In the original algorithm there's a goto b3 here which means
+        # the program will terminate, and thus return an empty list.
+        # But this is wrong, an example is partitions(1, 1, [1], [1]).
+        # What we need to do instead here is to break the loop.
+        break
       end
       j = j + 1
       k = mu[j]
       ll = v[j]
     end
+    if i == 1 
+      break
+    else
+      i = i - 1
+    end
   end #for i
-
+  
   lr = v[r]
   ll = v[1]
 
+  # This is a necessary condition for existence of a partition
   if m < 0 || m > n * (lr - ll)
-    return P
+    return P #goto b3
   end
 
-  if m == 0
-    push!(P,Partition{T}(x[1:n]))
+  # The following is a condition for when only a single partition
+  # exists. We added the || i==1 condition because without it the example
+  # partitions(17, 7, [1,4], [1,4]) returns [0, 0, 4, 4, 4, 4, 1], which is 
+  # nonsense. So, we have to make sure that all entries of x were modified in
+  # the initial backtracking, which means i was counted down to 1.
+  # Noticed on Mar 23, 2023.
+  if m == 0 && x[1] != 0
+    push!(P,Partition{T}(copy(x)))
     return P
   end
 
+  # Now, the actual algorithm starts
   i = 1
   m = m + y[1]
 
+  # label b1
   while gotob1 == true
     if !gotob2
       for j = mu[r]:-1:1
-        if m<=lr
+        if m<=lr 
           gotob2 = true
           break
         end
         x[i] = lr
         ii[i] = r - 1
+        if i==n # Added, otherwise get out of bounds
```

Comment:
Nitpick, not required to change for a merge:

Maybe like this? (I don't mind much, it's just that the surrounding code often seems to use it with spaces... this comment of course also applies to other parts of this PR)
```suggestion
        if i == n # Added, otherwise get out of bounds
```

Reply:
Taken care of in d902a23.